syntax = "proto3";

package podcaster;

enum DownloadStatus {
  NOT_DOWNLOADED = 0;
  DOWNLOAD_IN_PROGRESS = 1;
  DOWNLOAD_ERROR = 2;
  DOWNLOAD_SUCCESS = 3;
}

enum PlaybackStatus {
  NOT_PLAYING = 0;
  PLAYING = 1;
  PAUSED = 2;
}

message DownloadProgress {
  int32 downloaded_bytes = 1;
  int32 total_bytes = 2;
}

message PlaybackProgress {
  int32 elapsed_ms = 1;
  int32 total_ms = 2;
}

message Episode {
  string episode_uri = 1;
  string title = 2;
  string description_short = 3;
  string description_long = 4;
  DownloadStatus download_status = 5;
  DownloadProgress download_progress = 6;
  PlaybackStatus playback_status = 7;
  PlaybackProgress playback_progress = 8;
}

message Podcast {
  string podcast_uri = 1;
  string title = 2;
  repeated Episode episodes = 3;
}

message Empty {}

message DatabaseState {
  repeated Podcast podcasts = 1;
};

message EpisodeUri {
  string podcast_uri = 1;
  string episode_uri = 2;
}

message EpisodeUpdate {
  EpisodeUri uri = 1;
  oneof status {
    DownloadStatus new_download_status = 3;
    DownloadProgress new_download_progress = 4;
    PlaybackStatus new_playback_status = 5;
    int32 new_playback_progress = 6;
    int32 new_playback_duration = 7;
  }
}

service Podcaster {
  rpc State (Empty) returns (DatabaseState) {}
  rpc Refresh (Empty) returns (DatabaseState) {}
  rpc EpisodeUpdates (Empty) returns (stream EpisodeUpdate) {}
  rpc Download (EpisodeUri) returns (Empty) {}
  rpc CancelDownload (EpisodeUri) returns (Empty) {}
  rpc Play (EpisodeUri) returns (Empty) {}
  rpc Pause (EpisodeUri) returns (Empty) {}
  rpc Resume (EpisodeUri) returns (Empty) {}
  rpc Stop (EpisodeUri) returns (Empty) {}
  rpc Delete (EpisodeUri) returns (Empty) {} 
  rpc ShutdownIfNotPlaying (Empty) returns (Empty) {}
}

message Config {
  repeated string feed = 1;
};